pageserver/src/walredo.rs

```diff
@@ -727,27 +736,32 @@ impl PostgresRedoProcess {
         // all fallible operations post-spawn are complete, so get rid of the guard
         let child = scopeguard::ScopeGuard::into_inner(child);
 
-        Ok(PostgresRedoProcess {
-            tenant_id,
+        **input = Some(ProcessInput {
             child,
-            stdin,
-            stdout,
-            stderr,
-        })
-    }
+            stdin_fd: stdin.as_raw_fd(),
+            stdout_fd: stdout.as_raw_fd(),
+            stderr_fd: stderr.as_raw_fd(),
+            stream: stdin,
+            n_requests: 0,
+        });
+
+        *self.stdout.lock().unwrap() = Some(ProcessOutput {
+            stream: stdout,
+            pending_responses: VecDeque::new(),
+            n_processed_responses: 0,
+        });
+        *self.stderr.lock().unwrap() = Some(stderr);
```

Comment:
Since we no longer have a single mutex guarding these, I think it's possible to have threads T1, T2, T3:

1. T1 completes write
2. T2 completes write
3. T1 polls error while reading, bubbles error up, kills the child
4. T3 comes along starts the child
5. T3 completes write
6. T2 polls old fds (from ProcessOutput), acts on new ChildStd(out|err) (from mutexes), doesn't get error

Or am I missing something, apart from obviously very lucky scheduling?

Reply:
Trying not to do (renumbering) edits. I think T2 would in fact ingest result for T3 and place it on the vecdeque properly, but then await until timeout for the second response which it thinks it's theirs.

If there's a new request, then the response to that would be mixed up with T2's response, freeing mutex and letting T3 pick up the correct response, while T4 is left waiting until timeout (or new request from a new thread T5).

Perhaps I am missing some case where an overflow would happen and indexing into the vecdeque would panic.

Reply:
I have added comparison of output file descriptors and comment explaining why it is safe.

Reply:
I did some testing, it'd appear that at least on linux 5.15 fd's are allocated using some setup which will cause duplicates if `pipe` and `close` is called in loop... I don't know anything else about the allocation strategy. Do we have enough fd-churn to make this a valid strategy?

Generation based idea came to mind once more, either via `input: Mutex<(usize, Option<ProcessInput>)>` or a global atomicusize.